File: Python/ExampleTests/production/SimpleText.py
import re

_PATTERN = re.compile(r'[ a-zA-Z0-9,.\-]*')


class SimpleText:
    """Alphabetic, numeric, space, hyphen, period, comma — nothing else."""

    def __init__(self, value):
        self.value = value if value is not None else ''
        if not _PATTERN.fullmatch(self.value):
            raise ValueError(f'Invalid SimpleText: {self.value}')

    def __eq__(self, other):
        if not isinstance(other, SimpleText):
            return NotImplemented
        return self.value == other.value

    def __hash__(self):
        return hash(self.value)

    def __str__(self):
        return self.value

File: Rust/ExampleTests/production/simpletext.rs
/// Alphabetic, numeric, space, hyphen, period, comma — nothing else.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct SimpleText {
    pub value: String,
}

impl SimpleText {
    pub fn parse(value: &str) -> Result<Self, String> {
        let allowed = |c: char| c.is_ascii_alphanumeric()
            || matches!(c, ' ' | ',' | '.' | '-');
        if !value.chars().all(allowed) {
            return Err(format!("Invalid SimpleText: {value}"));
        }
        Ok(SimpleText { value: value.to_string() })
    }

    pub fn new(value: &str) -> Self { SimpleText { value: value.to_string() } }
}

impl std::fmt::Display for SimpleText {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.value)
    }
}
